docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+Modules are the core of Flax. This is a practical guide to the design and behavior of the `Module` class. By the end, you should feel comfortable to go off the beaten track and use Modules in new ways.
+
+
+## Overview
+
+### Definition
+
+Let’s start with a high-level overview of the Module lifecycle. We start by defining a simple Module:
+
+
+```python
+class MLP(nn.Module):
+  # 1. Attribute annotations
+  hidden_size: int
+  out_size: int
+
+  # 2. The `setup` method
+  def setup(self):
+    self.hidden = nn.Dense(self.hidden_size)
+    self.out = nn.Dense(self.out_size)
+
+  # 3. User methods  
+  def __call__(self, x):
+    a = self.hidden(x)
+    h = nn.relu(a)
+    return self.out(h)
+```
+
+
+This Module consists of:
+
+
+1. **_Attribute annotations_**, defined as [dataclass](https://docs.python.org/3/library/dataclasses.html) fields. These annotations automatically define a constructor.
+2. **The `setup` method,** that creates submodules and assigns them to attributes.
+3. A **user-defined method**. By convention, most Modules have just one  `__call__` method, but you can define multiple methods or use different method names.
+
+### Construction/initialization
+
+Now we want to construct and use the `MLP` Module:
+
+
+```python
+mlp = MLP(5, 3)
+variables = mlp.init(random.PRNGKey(0), x)
+y = mlp.apply(variables, x)
+```
+
+
+First, we construct an instance of `MLP` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The `MLP` constructor does not actually create variables or any internal state whatsoever. It’s best to think of it as a specification or template of the Module that contains functionality but no data.
+
+Let’s first take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling `init` is just a special case of `apply` and we could also code it as:
+
+
+```python
+# equivalent to: variables = mlp.init(random.PRNGKey(0), x)
+_, variables = mlp.apply({}, x, rngs={"params": random.PRNGKey(0)}, mutable=True)
+```
+
+
+Thus, `init` is nothing more than a special case of `apply` where:
+
+1. We call a module without any initial variables (an empty dict)
+2. A PRNG generator named “params” is always passed for randomly initializing parameters.
+3. All variable collections are set to mutable (`mutable=True`). When a collection is mutable, existing variables can be updated and new variables can be created.
+
+### Lifecycle
+
+Thus, what remains is to actually understand what `.apply(...)` does. All the interesting stuff in a Module happens inside the `apply` method. We will refer to the various stages before and during `apply` as the “Module lifecycle”. We can summarize this lifecycle in a number of steps:
+
+
+1. We construct `mlp = MLP(5, 3)`, such that `mlp.hidden_size=5` and `mlp_out_size=3`
```

Comment:
This is not really part of the `apply` function right? It seels like 2.1, 2.2, 2.3 below are actually the `apply` lifecycle.